### ISSUE TITLE
tf.constant() processes float16 100 times slower than float32

### ISSUE THREAD

**Issue by OliverPaulson** [tensorflow/tensorflow#19180]
```python
import tensorflow as tf
import numpy as np
import time

images = tf.keras.datasets.mnist.load_data()[0][0]

imagesFloat16 = images.astype(np.float16);
imagesFloat32 = images.astype(np.float32);

start = time.time()
tf.data.Dataset.from_tensor_slices(imagesFloat32)
end = time.time()
print("float32: {0:.4f} sec".format(end - start)) # float32: 1.2561 sec

start = time.time()
tf.data.Dataset.from_tensor_slices(imagesFloat16)
end = time.time()
print("float16: {0:.4f} sec".format(end - start)) # float16: 110.0273 sec
```

> float32: 1.2561 sec
> float16: 110.0273 sec

**Comment by mrry** [comment]
The `Dataset.from_tensor_slices()` method should mostly be a wrapper for `tf.constant()` (via `tf.convert_to_tensor()`). Can you please confirm if the following program has the same behavior?

```python
import tensorflow as tf
import numpy as np
import time

images = tf.keras.datasets.mnist.load_data()[0][0]

imagesFloat16 = images.astype(np.float16);
imagesFloat32 = images.astype(np.float32);

start = time.time()
tf.constant(imagesFloat32)
end = time.time()
print("float32: {0:.4f} sec".format(end - start)) # float32: 1.2561 sec

start = time.time()
tf.constant(imagesFloat16)
end = time.time()
print("float16: {0:.4f} sec".format(end - start)) # float16: 110.0273 sec
```

**Comment by mrry** [comment]
When I ran the version with `tf.constant()`, I got the following results:

> float32: 0.5304 sec
> float16: 113.5471 sec

...so I'll update the bug title to reflect that.

**Comment by mrry** [comment]
Not my part of the code, so I'm opening this up to contributions from within or outwith the team. Looks like somebody would need to write a function called `fast_tensor_util.AppendFloat16ArrayToTensorProto()` in [this file](https://github.com/tensorflow/tensorflow/blob/f08f24cd559b5824a1874a0e76d339875e43f366/tensorflow/python/framework/fast_tensor_util.pyx) and call it [from here](https://github.com/tensorflow/tensorflow/blob/f08f24cd559b5824a1874a0e76d339875e43f366/tensorflow/python/framework/tensor_util.py#L71).

**Comment by faintedremix** [comment]
Mind if I take this case? :)

**Comment by mrry** [comment]
@faintedremix We’d appreciate it!

**Comment by facaiy** [comment]
@faintedremix Sorry that I have wrote a PR #19212  before I see your comment. However, my PR can't fix the problem totally, you can take over it  if you have a better solution :-)

**Comment by yongtang** [comment]
Ah I didn't check the comment when creating the PR #19213. Let me close the PR #19213 for now.

**Comment by pbanavara** [comment]
@faintedremix Have you taken this up and started working ? If not, please let me know I'm happy to take over or if you have opened a PR, I'm happy to pitch in and help

**Comment by yongtang** [comment]
@pbanavara The issue has been fixed in PR #19212 I think. Also there is another PR #19533 that is pending which I think will address the similar `bfloat16` issue.

**Comment by karmel** [comment]
Closing as the fixing PR is merged. If that is incorrect, please reopen.
